oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentRevisionCleaner.java

```diff
@@ -0,0 +1,270 @@
+
+            this.revisionsModifyingPropertyByCluster = new TreeMap<>();
+            this.revisionsModifyingProperty = new TreeMap<>();
+            this.propertiesModifiedByRevision = new TreeMap<>(StableRevisionComparator.INSTANCE);
+        }
+
+        /**
+         * This method processes the revisions of the working document, creating maps to
+         * track the relationships between revisions and modified properties.
+         */
+        protected void classifyRevisionsAndProperties() {
+            SortedMap<Revision, String> deletedRevisions = workingDocument.getLocalDeleted();
+            // Always keep the first "_deleted" entry, as is when the document was created
+            if (!deletedRevisions.isEmpty()) {
+                Revision createdRevision = deletedRevisions.firstKey();
+                revisionCleaner.addBlockedRevisionToKeep(createdRevision);
+            }
+
+            SortedMap<Revision, String> documentRevisions = workingDocument.getLocalRevisions();
+            for (Map.Entry<Revision, String> revisionEntry : documentRevisions.entrySet()) {
+                Revision revision = revisionEntry.getKey();
+                String revisionValue = revisionEntry.getValue();
+
+                // Only check committed revisions
+                if (Utils.isCommitted(revisionValue)) {
+                    // Candidate to clean up
+                    revisionCleaner.addCandidateRevisionToClean(revision);
+                    // Store properties usage
+                    classifyPropertiesModifiedByRevision(revision);
+                }
+            }
+        }
+
+        private void classifyPropertiesModifiedByRevision(Revision revision) {
+            for (Map.Entry<String, Object> propertyEntry : workingDocument.entrySet()) {
+                if (Utils.isPropertyName(propertyEntry.getKey()) || NodeDocument.isDeletedEntry(propertyEntry.getKey())) {
+                    Map<Revision, String> valueMap = (Map) propertyEntry.getValue();
+                    if (valueMap.containsKey(revision)) {
+                        propertiesModifiedByRevision.computeIfAbsent(revision, key ->
+                                new TreeSet<>()).add(propertyEntry.getKey()
+                        );
+
+                        revisionsModifyingPropertyByCluster.computeIfAbsent(propertyEntry.getKey(), key ->
+                                new TreeMap<>()
+                        ).computeIfAbsent(revision.getClusterId(), key ->
+                                new TreeSet<>(StableRevisionComparator.INSTANCE)
+                        ).add(revision);
+
+                        revisionsModifyingProperty.computeIfAbsent(propertyEntry.getKey(), key ->
+                                new TreeSet<>(StableRevisionComparator.INSTANCE)
+                        ).add(revision);
+                    }
+                }
+            }
+        }
+    }
+
+    protected class RevisionCleanerUtility {
```

Comment:
Is there a reason this is not `private` ?

Reply:
Left over from initial testing. I have made it private.